cmd/storj/cmd/cp.go

```diff
@@ -32,79 +37,164 @@ func init() {
+	// if object name not specified, default to filename
+	if destObj.Path == "" || destObj.Path == "/" {
+		destObj.Path = filepath.Base(srcFile)
 	}
 
-	if len(args) == 1 {
-		return errs.New("No destination specified")
+	f, err := os.Open(srcFile)
+	if err != nil {
+		return err
 	}
 
-	so, err := getStorjObjects(ctx, cpCfg)
+	defer utils.LogClose(f)
+
+	o, err := bs.GetObjectStore(ctx, destObj.Host)
 	if err != nil {
 		return err
 	}
 
-	u, err := url.Parse(args[0])
+	meta := objects.SerializableMeta{}
+	expTime := time.Time{}
+
+	_, err = o.Put(ctx, paths.New(destObj.Path), f, meta, expTime)
 	if err != nil {
 		return err
 	}
 
-	if u.Scheme == "" {
-		f, err := os.Open(args[0])
-		if err != nil {
-			return err
-		}
+	fmt.Printf("Created: %s\n", destObj.Path)
 
-		fi, err := f.Stat()
-		if err != nil {
-			return err
-		}
+	return nil
+}
 
-		fr, err := hash.NewReader(f, fi.Size(), "", "")
-		if err != nil {
-			return err
-		}
+// download downloads s3 compatible object args[0] to args[1] on local machine
+func download(ctx context.Context, bs buckets.Store, srcObj *url.URL, destFile string) error {
+	o, err := bs.GetObjectStore(ctx, srcObj.Host)
+	if err != nil {
+		return err
+	}
 
-		defer func() { _ = f.Close() }()
+	f, err := os.Create(destFile)
+	if err != nil {
+		return err
+	}
 
-		u, err = url.Parse(args[1])
-		if err != nil {
-			return err
-		}
+	defer utils.LogClose(f)
 
-		oi, err := so.PutObject(ctx, u.Host, u.Path, fr, nil)
-		if err != nil {
-			return err
-		}
+	rr, _, err := o.Get(ctx, paths.New(srcObj.Path))
+	if err != nil {
+		return err
+	}
+	defer utils.LogClose(rr)
 
-		fmt.Println("Bucket:", oi.Bucket)
-		fmt.Println("Object:", oi.Name)
+	r, err := rr.Range(ctx, 0, rr.Size())
+	if err != nil {
+		return err
+	}
+	defer utils.LogClose(r)
 
-		return nil
+	_, err = io.Copy(f, r)
+	if err != nil {
+		return err
 	}
 
-	oi, err := so.GetObjectInfo(ctx, u.Host, u.Path)
+	fmt.Printf("Downloaded %s to %s\n", srcObj.Path, destFile)
+
+	return nil
+}
+
+// copy copies s3 compatible object args[0] to s3 compatible object args[1]
+func copy(ctx context.Context, bs buckets.Store, srcObj *url.URL, destObj *url.URL) error {
+	o, err := bs.GetObjectStore(ctx, srcObj.Host)
 	if err != nil {
 		return err
 	}
 
-	f, err := os.Create(args[1])
+	rr, _, err := o.Get(ctx, paths.New(srcObj.Path))
 	if err != nil {
 		return err
 	}
+	defer utils.LogClose(rr)
 
-	defer func() { _ = f.Close() }()
+	r, err := rr.Range(ctx, 0, rr.Size())
+	if err != nil {
+		return err
+	}
+	defer utils.LogClose(r)
 
-	err = so.GetObject(ctx, oi.Bucket, oi.Name, 0, oi.Size, f, oi.ETag)
+	o, err = bs.GetObjectStore(ctx, destObj.Host)
 	if err != nil {
 		return err
 	}
 
-	fmt.Printf("Downloaded %s to %s", oi.Bucket+oi.Name, args[1])
+	meta := objects.SerializableMeta{}
+	expTime := time.Time{}
+
+	// if destination object name not specified, default to source object name
+	if destObj.Path == "" || destObj.Path == "/" {
```

Comment:
add the filename if destObj.Path == "/my/folder/prefix/thing/" too? (like before)

Reply:
Done